composition-js/src/merging/merge.ts

```diff
@@ -356,13 +374,13 @@ class Merger {
     // Then, for object and interface types, we merge the 'implements' relationship, and we merge the unions.
     // We do this first because being able to know if a type is a subtype of another one (which relies on those
     // 2 things) is used when merging fields.
-    for (const objectType of this.merged.types<ObjectType>('ObjectType')) {
+    for (const objectType of this.typesToMerge<ObjectType>('ObjectType')) {
```

Comment:
These should probably be separate functions. `this.typesToMerge<ObjectType>('InterfaceType');` is valid code, but could cause typing errors.

Reply:
Oh yeah, the fact that we're casting inside of `Schema::types` before returning is definitely a code smell.

Reply:
> `this.typesToMerge<ObjectType>('InterfaceType');` is valid code

I'll note that this isn't really true, that code did *not* compile (the `kind` argument _has_ to match the kind of the type parameter because it's type is `T['kind']`).

That said, that method (and `schema::types`) was indeed a bit error prone as `this.typesToMerge<ObjectType>()` _was_ valid code, yet didn't return only the object types. I initially went with it as I figured it would obvious that simply passing a type argument couldn't influence runtime behaviour and it would obvious you should do this, but admittedly not my best moment in hindsight.

Anyway, I cleaned this up a bit to avoid that potential problem, but the code in `merge.ts` still uses a single `filterTypesOfKind<ObjectType>(typesToMerge, 'ObjectType')` method because it _is_ type safe now (the `kind` argument is not optional anymore) and it didn't felt cleaner adding a bunch of single-usage methods.